jsparagus/aps.py

```diff
@@ -190,44 +209,62 @@ def shift_next(self, pt: ParseTable) -> typing.Iterator[APS]:
         st, sh, la, rp, hs = self.stack, self.shift, self.lookahead, self.replay, self.history
         last_edge = sh[-1]
         state = pt.states[last_edge.src]
-        state_match_shift_end = True
+        state_match_shift_end = self.state == self.shift[-1].src
         if self.replay == []:
+            assert state_match_shift_end
             for term, to in state.shifted_edges():
-                edge = Edge(last_edge.src, term)
+                edge = Edge(self.state, term)
                 new_sh = self.shift[:-1] + [edge]
                 edge_to = Edge(to, None)
-                yield APS(st, new_sh + [edge_to], la + [term], rp, hs + [edge], False)
-        else:
+                yield APS(to, st, new_sh + [edge_to], la + [term], rp, hs + [edge], False)
+        elif state_match_shift_end:
             term = self.replay[0]
             rp = self.replay[1:]
             if term in state:
-                edge = Edge(last_edge.src, term)
+                edge = Edge(self.state, term)
                 new_sh = self.shift[:-1] + [edge]
                 to = state[term]
                 edge_to = Edge(to, None)
-                yield APS(st, new_sh + [edge_to], la, rp, hs + [edge], False)
+                yield APS(to, st, new_sh + [edge_to], la, rp, hs + [edge], False)
 
         rp = self.replay
         for a, to in state.epsilon:
-            edge = Edge(last_edge.src, a)
+            edge = Edge(self.state, a)
             prev_sh = self.shift[:-1] + [edge]
             # TODO: Add support for Lookahead and flag manipulation rules, as
             # both of these would invalide potential reduce paths.
             if a.update_stack():
+                stack_diff = a.update_stack_with()
+                if stack_diff.replay < 0:
+                    assert stack_diff.pop == 0
+                    assert stack_diff.nt is None
+                    num_replay = -stack_diff.replay
+                    assert len(self.replay) >= num_replay
+                    new_rp = self.replay[:]
+                    new_sh = self.shift[:]
+                    while num_replay > 0:
+                        num_replay -= 1
+                        term = new_rp[0]
+                        del new_rp[0]
+                        sh_state = new_sh[-1].src
+                        sh_edge = Edge(sh_state, term)
+                        sh_to = pt.states[sh_state][term]
+                        sh_edge_to = Edge(sh_to, None)
+                        del new_sh[-1]
+                        new_sh = new_sh + [sh_edge, sh_edge_to]
```

Comment:
It seems like this loop should either check or assert that we won't reach a state where `pt.states[sh_state][term]` is an action state.

Reply:
No, actually we can and expect so. This expected when the 2 non-terminal have the same number of lookahead.

For Example:
```
  Add <-- Num `+` Mul [lookahead: X Y]
  Mul <-- Num `+` Num [lookahead: X Y]
```

In this example, replacing the `Reduce` of `Mul` non-terminal would be replaced by `Unwind`, `FilterState` and `Replay`, and the Replay instruction would replay `Mul`, `X`, `Y` terms and reach the `sh_to` state of the `Reduce` of `Add` (which will also be replaced by `Unwind`, `FilterState` and `Replay`)

The assertion you suggest is only valid for `pt.states[sh_state]`, but there is no need to assert anything as an `IndexError` should occur if the term is not present as an outgoing edge.